### DISCUSSION TITLE
What would be the best way to make an AI environment using ruffle?

### DISCUSSION THREAD

**Discussion by Gumbo64** [ruffle-rs/ruffle#7005] [Q&A]
I'm aiming to make an AI to play spiderman city raid, which will need mouse input. 

I've been thinking about:
1. use web ruffle with selenium
- Can use the pause/play functions of the API to play frame by frame (it places a huge play button on the screen though, anyone know how to get rid of that?)
- Tried dispatchEvent and .click(), didn't work. Not sure if python selenium / javascript functions can use mouse input on ruffle

2. use desktop ruffle with some homebrew/modding
- doesn't have API (I think?)
- maybe something similar to the --timedemo option with uncapped framerate
- maybe with libTAS although I don't know much about it (https://tasvideos.org/EmulatorResources/Ruffle)
- Any tips on how to connect ruffle desktop or libTAS to python?

Because its an emulator, I'm wondering if it has a way to input straight into the game without stuff like pyautogui which can be messy. Also playing frame-by-frame would be good but not a necessity

**Comment by danielhjacobs** [comment]
I don't know much, but to answer one small question:

> it places a huge play button on the screen though, anyone know how to get rid of that?

Assuming your ruffle-player has id player, right after calling the pause function, do `document.getElementById("player").shadowRoot.getElementById("play_button").style.display = "none";` See https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot. Inherits from https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment, which contains the getElementById, querySelector, and querySelectorAll functions.

Rythin is in the Ruffle Discord server so you can try reaching out to her. She TASed Flash Meat Boy with Ruffle and libTAS: https://www.youtube.com/watch?v=MUhwSHTvAn4. Following the instructions you linked for using Ruffle with libTAS is probably your best bet, as it proved to work for others.

**Reply by Gumbo64** [reply]
Niice i got selenium to work with the mouse clicks too so the web version is definitely my best bet at the moment. Turns out when you pause the player, you can click on the screen (with play button removed) and it will click there ingame once you unpause, which is just like input buffering in TAS. With that stuff I just made a js button for selenium to press that pauses and unpauses after 30ms (~1 frame). I think that should be good enough for my purposes so thanks for helping out!

code for future people
```javascript

function pausegame(){
    // gets the <ruffle-player> element in html
    let get_player = document.querySelector('#gamers > ruffle-player:nth-child(1)')
    get_player.pause()
    // overwrites style to none after .pause() updates it
    get_player.shadowRoot.getElementById("play_button").style.display = "none";
  }
function playgame(){
    let get_player = document.querySelector('#gamers > ruffle-player:nth-child(1)')
    get_player.play()
    get_player.shadowRoot.getElementById("play_button").style.display = "none";
  }
function frameforward(){
    playgame()
    setTimeout(pausegame, 30);
  }
```
also python selenium you just use something like:
```python
# element ID to 'anchor' from, just use <body> if you have the player in the top left corner like me 
el=driver.findElement(By.ID,"**ID**")
action = webdriver.common.action_chains.ActionChains(driver)
action.move_to_element_with_offset(el, 100, 100)
action.click()
action.perform()
```
also I think if you hover over the ruffle player while selenium is running then your mouse position overrides selenium's so watch out for that


edit: decompiling with jpexs + http requests (LoadVars) is what I ended up using with the desktop version. ruffle's flashvar launch commands are also crucial to set up different sockets for playing with multiple games at the same time (multi agent)
